### ISSUE TITLE
O(n)-sized jaxpr takes O(exp(n)) work to obtain

### ISSUE THREAD

**Issue by patrick-kidger** [google/jax#8184]
The following code has exponential compilation time in the depth (here passed as `10`):

```python
import jax
import jax.lax as lax
import jax.numpy as jnp

def fn(x, depth):
    if depth == 1:
        return x
    else:
        _fn = lambda x: fn(x, depth - 1)
        return lax.cond(jnp.array(True), _fn, lambda x: x, x)

jax.make_jaxpr(jax.vmap(lambda x: fn(x, 10)))(jnp.array([1.]))
```

The reason is that the condition `jnp.array(True)` is not vmap'd, which means that two calls to `batch_jaxpr` are made:

Here:
https://github.com/google/jax/blob/d8b7bd54be7252673a211b8f4d06a0b4d72256c1/jax/_src/lax/control_flow.py#L857
and here:
https://github.com/google/jax/blob/d8b7bd54be7252673a211b8f4d06a0b4d72256c1/jax/_src/lax/control_flow.py#L861
So that overall `2^depth` calls are made to `batch_jaxpr`.

I'm currently writing something involving a binary tree, but unfortunately this is pretty much a blocker.

**Comment by patrick-kidger** [comment]
I've edited my original comment -- the original had a typo that meant it did indeed have an exponentially-sized jaxpr. This one has a linearly-sized jaxpr with exponential compile time.

**Comment by shoyer** [comment]
Others may have ideas for how to do this, but unfortunately, I'm not sure if it's possible to do tree traversal efficiently in XLA.

**Comment by patrick-kidger** [comment]
I'm only calling `make_jaxpr` above, so I don't think XLA even comes into it. It's just a linear-size jaxpr that takes exponential time to obtain.

**Comment by patrick-kidger** [comment]
Coming back to this old issue: this arose whilst implementing [Diffrax](https://github.com/patrick-kidger/diffrax). (A new suite of differential equation solvers for JAX.)

The workaround for this issue (and #8193, which is of similar character) was to monkey-patch the JAX interpreter, [here](https://github.com/patrick-kidger/diffrax/blob/2b4e4d863c15abc7143919bac7825090bbfe50be/diffrax/misc/bounded_while_loop.py#L248).

**Comment by mattjj** [comment]
I was just looking at making our fixpoints asymptotically faster, and I noticed that indeed the repro in the OP is actually already fixed at HEAD, probably because of the caching already discussed (finally added in #9196 and #9228). That is, it's fast to execute now (even upping the constant all the way until hitting a Python recursion limit error), but I'm not sure if it's got the right asymptotics (we might be hitting a cache exponentially many times, rather than not doing any exponential work at all).

I'll look for a slow repro to work on, but just wanted to update the thread in the meantime.

**Comment by mattjj** [comment]
Hrm really doesn't seem exponential anymore:
<img width="1270" alt="image" src="https://user-images.githubusercontent.com/1458824/156850666-c66f0574-ffd7-413d-bd07-e997cba9ee62.png">

**Comment by mattjj** [comment]
Okay here's a hard one:

```python
from functools import partial
import jax
from jax import lax
import jax.numpy as jnp


def f(n, x):
  def body(_, xs):
    x, *rest = xs
    x = f(n - 1, x) if n != 0 else x
    return [*rest, x]

  xs = [jnp.zeros_like(x)] * 100
  xs[0] = x
  x, *_ = lax.fori_loop(0, jnp.array(100), body, xs)
  return x

jax.make_jaxpr(jax.vmap(partial(f, 3)))(jnp.array([0]))
```

**Comment by mattjj** [comment]
@dougalm found an exponential case, even with caching! (After I tried and failed to find one.) **The text below is just quoting his chat messages.**

Consider:
```
\x1.
  loop 0 \x2 .
    loop 0 \x3 .
      loop 0 \x4 .
        x1
      x1
    x1
  x1
```

Or if we make carries explicit:

```
\x1.
  loop x1 0 \_ x2 .
    loop x1 x2 0 \_ _ x3 .
      loop x1 x2 x3 0 \_ _ _ x4 .
        (0, 0, 0, x1)
      (0, 0, x1)
    (0, x1)
  x1
```

It goes like this:
1. `x1` is always `1`
2. you start the outer loop with `10`
3. then you go to the next loop with `100`, and so on
4. eventually you bottom out in the innermost loop with `1000`, which after running its body produces `1001`
5. so you try again with `1001`, and hooray that's a fixed point! so you've done two passes of the innermost loop, and you have sweet cache entries for `1000` and `1001`
6. now you return to the next outer loop, which you had been querying with `100`, and you rerun again with `101`
7. that forces you to try the innermost loop again with `1010`, and _oh no, missed the cache!_

We can transcribe that directly to JAX as something like

```python
def f(x1):
  def loop(xs):
    _, x2 = xs
    def loop(xs):
      _, _, x3 = xs
      def loop(xs):
        _, _, _, x4 = xs
        return (0, 0, 0, x1)
      while_loop(lambda _: False, loop, (x1, x2, x3, 0))
      return (0, 0, x1)
    while_loop(lambda _: False, loop, (x1, x2, 0))
    return (0, x1)
  while_loop(lambda _: False, loop, (x1, 0))
  return x1
```

Or if we want to parameterize the depth:

```python
def f(xs):
  if len(xs) < depth:
    while_loop(lambda _: False, f, (*xs, 0))
  return (0,) * (len(xs) - 1) + xs[:1]
```

Indeed running that last program, even with slightly smarter caching that @apaszke and I just noticed we needed, gives exponential number of jaxpr passes as a function of `depth`.

@dougalm proposed a fix (similar ideas have come up before from Adam in this thread and @hawkinsp), but he/we don't know if it's worth the power:weight. We want to try some real-world examples.

@patrick-kidger do you have any real-world examples (e.g. from diffrax) where JAX spends too much time in these fixed point computations?

**Comment by patrick-kidger** [comment]
Hmm. So you've successfully nerd-sniped me in to trying to give a quantitative answer, but the end result is "I'm not sure".

On one hand: at least in terms of asymptotics, the caching does seem to fix this issue when tracing Diffrax programs. In Diffrax the nested scans (and nested conds) have an argument with the same structure at every level. I think that means there's no room for any kind of "vmap demotion" a la the final `1 -> 0` in `1001 -> 1010` in your example, so things run in linear time. Hurrah!

Conversely, it may still be possible to improve by a meaningful constant factor. For example if we could halve compile times that would still be a big win. If we pretend that discerning input/output dependencies is free, and observe that `lax.cond` currently requires a length-2 fixed-point iteration, then we'd halve the computational work here -- and that would be amazing. Likewise we would see larger gains on loops. I've now measured what are typical numbers of fixed-point iterations for the loops used throughout Diffrax. Mostly it's a mix of 1, 2, 3 and the occasional 4.

Either way -- if you decide to try something in this vein then the examples of Diffrax are probably good benchmarks to try, given the heavy nesting of scans and conds in `diffrax.diffeqsolve`.

**Comment by mattjj** [comment]
@patrick-kidger thanks for those thoughts. Good to hear that the exponential time behavior is fixed for your use cases. I agree there's more improvement to be done, including constant factors.

One tricky aspect of computing the dependence once up-front, then doing iterations for which the cost is independent of the jaxpr size (i.e. number of eqns), is that 'dependence' is different for every transformation, and indeed different on a primitive-by-primitive basis. For example, when computing a fixed point for a AD on a `while_loop` using `body_fun` below, we don't want to count the second output as "dependent" on the first input (in the sense that the second output does not need a corresponding tangent value to be computed):

```python
def f(x: float, b: bool):
  return x, x == 0
```

Here's an example for `vmap`, if we consider `x` to be mapped over the axis name `i`:

```python
def f(x, y):
  return x, lax.psum(x, 'i')
```

That's not a fundamental difficulty, and we have a prototype which takes these difference into account. But it does affect the power-to-weight ratio of the solution, and hence may affect whether the added complexity is worthwhile.

I'll investigate how much faster we might make the diffrax examples. I'm sure I could figure this out, but in case it's easy, could you tell me the best way to run them (or which to focus on)?

**Comment by patrick-kidger** [comment]
Ach. I hadn't thought about the fact that 'IO dependence' is itself transformation-dependent. How does the prototype work? (Of the form `@cache() def batch_dependence(jaxpr, ...)` or something else?)

In terms of the Diffrax examples -- no worries at all. They're available as notebooks [here](https://github.com/patrick-kidger/diffrax/tree/main/examples) and they're all runnable as standalone files: `ipython neural_ode.ipynb` etc. Obviously cut out the stuff you're not interested in timing. `stiff_ode.ipynb` is the simplest of them, and all of them except `symbolic_regression.py` should be useful as benchmarks.

---

Since we're discussing IO dependence -- something I'd love to see is a IO dependency tracker operating at the level of individual array elements, as part of the public API. This has cool applications like Jacobian sparsity detection, or checking certain conditions for higher-order diffeq solves. Now in practice this would clearly need a whole new interpreter (so as to handle each primitive appropriately) so my expectation is that this is way out of scope right now. But I thought I'd mention on the off-chance that it's relevant here.

**Comment by mattjj** [comment]
> How does the prototype work?

With tables of rules! Our favorite thing. That is, the default dependence rule is that all outputs of a primitive application depend on all inputs, and then we override that behavior for specific primitives for each transformation. So higher-order primitives (like xla_call, scan, etc) would track dependence in a more fine-grained way, and we'd also handle the above examples.

Luckily there are relatively few such exceptions, as far as I can tell. So there aren't many rules to add.

A downside to that approach is that if we forget to add a special rule for a transform/primitive pair, we might end up over-estimating the dependence and e.g. introduce extra batch dimensions (in the vmap case).

Another downside is just the complexity. It's not a lot of lines of code but it's not nothing either.

> something I'd love to see is a IO dependency tracker operating at the level of individual array elements, as part of the public API. [...] Now in practice this would clearly need a whole new interpreter (so as to handle each primitive appropriately) so my expectation is that this is way out of scope right now

Interpreters are pretty easy to stand up, but they're costly to maintain because they introduce more constraints or code-to-be-updated on the JAX core system. That is, they can slow down JAX core development. In other words, the main cost to be aware of is in "number of new APIs" rather than strictly in implementation complexity.

I'd be happy to help prototype something like that, but it's very unlikely we'd want to add it to the JAX API unless we find a lot of common/core use cases.

Thanks for the instructions on the diffrax examples!

**Comment by mattjj** [comment]
To say a bit more about the prototype (and I should probably just paste the code as it's pretty small): we have a `jaxpr_dependence(special_rules: Dict[Primitive, Callable], jaxpr: Jaxpr) -> List[Set[Var]]` function which computes how a jaxpr's outputs depend on its inputs. The `special_rules` argument represents transform- and primitive-specific rules (so `while_loop`'s batching rule calls `jaxpr_dependence` with a different dict from `while_loop`'s JVP rule). Then to compute fixed points we can use that to build a boolean matrix (in a single pass over the jaxpr) and then use repeated squaring to find a fixed-point of the iterated transfer function.

**Comment by patrick-kidger** [comment]
Haha, makes sense. Yeah, I'm not sure what a SSOT implementation looks like for this. Maybe an autogenerated table of rules that manually runs each interpreter-primitive pair, and checks the dependence, the first time each such pair is encountered? In particular I'm thinking about what happens when someone writes their own interpreter or their own primitive. If nothing else I'm invested in this as Diffrax has a few custom primitives.

On interpreters. Yup, this is pretty much what I thought you'd say. I thought I'd throw it out there juuuust in case you were going for an interpreter-based solution, in which case it would naturally be a "common/core use case".

As much for my own curiosity: I'm not sure what you mean by "repeated squaring". Naively I'd try and find a fixed point by just iterated multiplication of the dependency matrix until the result stabilises. But what you're saying makes me think of something like optimal [addition chains](https://en.wikipedia.org/wiki/Addition_chain) when exponentiating? (Or is that way off the mark.)

**Comment by mattjj** [comment]
> In particular I'm thinking about what happens when someone writes their own interpreter or their own primitive. 

Well, if you add your own primitive, you've gotta populate a bunch of rules tables. Nothing new there!

Would Diffrax primitives need special 'dependency' treatment? I think for batching, for example, only axis-name-collectives need special treatment (and higher-order primitives).

> As much for my own curiosity: I'm not sure what you mean by "repeated squaring". 

Oh just that if you want to compute lim_{n \to \infty} A^n iteratively, it's better to iterate X -> X^2 with X_0 = A rather than iterating X -> A X. The former will reach a fixed point exponentially faster.

**Comment by patrick-kidger** [comment]
> Well, if you add your own primitive, you've gotta populate a bunch of rules tables. Nothing new there!

Yeah, but if we can minimise the table-filling-out that would be great ;)

> Would Diffrax primitives need special 'dependency' treatment? I think for batching, for example, among JAX primitives only axis-name-collectives need special treatment (and higher-order primitives).

Some of them definitely would. There's a family of primitives like `unvmap_any` that perform reductions ignoring batch dimensions. [They're defined here.](https://github.com/patrick-kidger/diffrax/blob/3f4f5faadd3becd2099c4bda41f7dc448e4078f8/diffrax/misc/unvmap.py). (The other new primitive in Diffrax is one which approximates error-on-vjp by raising an error-on-transpose, [here](https://github.com/patrick-kidger/diffrax/blob/29372abce0a0f40750bfecfd7188fe9ea2e61c9e/diffrax/misc/ad.py#L31). But I don't think that one would need special treatement.)

FWIW I could probably reimplement both of these without new Primitives by using the custom batching and custom transposition functionality, once those are finished. (Could the dependency tracking system handle `unvmap_any` implemented in such a way?)

> Oh just that if you want to compute lim_{n \to \infty} A^n iteratively, it's better to iterate X -> X^2 with X_0 = A rather than iterating X -> A X. The former will reach a fixed point exponentially faster.

Right, I was wondering if you were up to something like this. But I can't see why that's the thing needed. Isn't the fixed point that's desired the solution of `min sum(v)` where `v \in {0, 1}^d`, subject to `Av = v` and `v_i = 1 for all i which are batched/jvp'd/etc.`?

It's not clear to me that `lim_n->inf A^n` is even defined. What if `A = [[0, 1], [1, 0]]`.

**Comment by mattjj** [comment]
> Isn't the fixed point that's desired the solution of argmin_v sum(v) where v \in {0, 1}^d, subject to Av = v and v_i = 1 for all i which are batched/jvp'd/whatever.

Yes but that's a problem statement, not an algorithm for solving it!

> It's not clear to me that lim_n->inf A^n is even defined. What if A = [[0, 1], [1, 0]].

Indeed, I was just describing what repeated squaring is.

Let A be the dependence matrix, so that A_{ij} = True if output i depends on input j and False otherwise. Let's assume for simplicity that it's square, and our while_loop has the same outputs as inputs (i.e. no "consts" as in JAX's actual while_loop, though those are easy to handle by assembling a block matrix). Then we want a fixed point of the map v -> A v + v (for boolean matrices/vectors and operations, where True + True = True). The addition corresponds to asking that the output be at least as batched (or whatever) as the input. That map can also be written as v -> (A + I)v, which is why we can use repeated squaring. (This is essentially a standard DAG transitive closure algorithm.)

I suspect that there are a lot of ways to think about the existence of the limit for this case (with the identity matrix / self-transitions added in): combinatoric, linear algebraic, random walks, etc.

The repeated squaring isn't important at all; it's just interesting to me from an asymptotics point of view. Of course there are instances where iterating matrix-vector multiplication is faster, but it depends on the connectivity of A.

**Comment by patrick-kidger** [comment]
Ah, I think I see the confusion; we've been talking past each other. I've been discussing "how do we find `v` for one particular problem, with known `v_i = 1` constraints".

Meanwhile, by precomputing the full transitive closure, you're able to find the solution `v`, for *any* set of constraints `v_i = 1`, by doing a matrix-vector product.

I think the naive iteration I was suggesting is optimal if you only need to solve the problem once. Meanwhile if you need to solve the problem many many times then precomputing the transitive closure is clearly optimal.

In the intermediate regime of solving the problem only a few times per jaxpr -- I think things could go either way. The naive iteration uses cheap matrix-vector products; computing the transitive closure via repeated squaring uses more expensive matrix-matrix products.
Conversely, the number of iterations to find the transitive closure of DAG may be reduced if the transitive closure of sub-DAGs are already known (which e.g. may be very common for the little sub-DAGs corresponding to `jnp.*` operations), and that's not a trick the naive iteration can take advantage of.

...that's a lot of words. Probably this is bikeshedding at this point. I'll leave it there! I think we understand each other.

**Comment by mattjj** [comment]
Actually I was talking about computing for a particular vector; that's why I was talking about the vector iteration v -> A v + v.

Surprisingly even if you want to compute the result for a single sparsity pattern it can be asymptotically faster to do the matrix iteration. It relies on the fact that matrix-matrix multiplication is o(N^3). Consider the connectivity matrix `A = np.eye(N, k=1, dtype='bool')` and the vector `v = np.array([True, False, False, False, ..., False])`. If you iterate `v -> A v + v` then it requires N-1 steps to converge with N^2 operations per step, for N^3 total operations. If you do iterated squaring then you need lg N steps each with complexity O(N^2.8) (with Strassen) or O(N^2.373) if you want a galactic algorithm, meaning O(N^2.8 * lg N) steps. That's asymptotically faster than N^3. Matrix multiplication is magic! For asymptotic analysis at least. That was the academically interesting asymptotics I was alluding to.

Actually, I think there are even better algorithms for doing this and the logarithmic factor can be removed. I'm not really familiar with them though. Repeated squaring is just a cheap and easy hack.

Practically speaking, we should just measure whichever seems faster in practice, and consider e.g. what we might want to cache/reuse.

**Comment by mattjj** [comment]
However I've been assuming dense `A`, while the concrete example I gave was for a sparse `A` for which matrix-vector multiplies could be O(N). I'm not sure if there's an example matrix for which a sparse matrix-vector iteration would be asymptotically slower than the dense matrix-matrix iteration...

In any case, asymptotic analysis won't tell us which approach is faster in practice (sparse matrix-vector, dense matrix-vector, dense matrix-matrix), so it always comes down to measurement.

**Comment by patrick-kidger** [comment]
~~Ah, yes -- I getcha. Although life gets even more interesting if one considers sparse matrix-matrix or matrix-vector multiplication algorithms~~ (You just spotted the same thing thing as I was writing it out!)

Yup, this is a can of worms. The possibility of caching makes life more complicated still. And I think it's clear that the best single solution is always dependent on the statistics of real-world programs.

So If you do actually experiment with this I'd be curious to know the results.

**Comment by mattjj** [comment]
I mentioned a very abstracted version of this problem to our colleague @poolio in the microkitchen, and he instantly came up with a solution: a chain with back-edges, so basically `A = np.triu(np.ones((N, N), dtype='bool'), k=-1)`. That has Theta(N^2) entries (so matrix-vector multiplication takes "dense time") yet in the worst case the vector iteration takes Theta(N) steps to converge (with `v = np.array([True] + [False] * (N-1))`). I think that's a matrix for which the matrix iteration is asymptotically faster. So cool.

**Comment by patrick-kidger** [comment]
Closing as I think this issue has been resolved for a while now.
